web/src/main/java/org/springframework/security/web/server/ServerFormPostRedirectStrategy.java

```diff
@@ -0,0 +1,124 @@
+
+	private static final StringKeyGenerator DEFAULT_NONCE_GENERATOR = new Base64StringKeyGenerator(
+			Base64.getUrlEncoder().withoutPadding(), 96);
+
+	private static final String REDIRECT_PAGE_TEMPLATE = """
+			<!DOCTYPE html>
+			<html lang="en">
+			  <head>
+			    <meta charset="utf-8">
+			    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
+			    <meta name="description" content="">
+			    <meta name="author" content="">
+			    <title>Redirect</title>
+			  </head>
+			  <body>
+			    <form id="redirect-form" method="POST" action="{{action}}">
+			      {{params}}
+			      <noscript>
+			        <p>JavaScript is not enabled for this page.</p>
+			        <button type="submit">Click to continue</button>
+			      </noscript>
+			    </form>
+			    <script nonce="{{nonce}}">
+			      document.getElementById("redirect-form").submit();
+			    </script>
+			  </body>
+			</html>
+			""";
+
+	private static final String HIDDEN_INPUT_TEMPLATE = """
+			<input name="{{name}}" type="hidden" value="{{value}}" />
+			""";
+
+	@Override
+	public Mono<Void> sendRedirect(ServerWebExchange exchange, URI location) {
+		String nonce = DEFAULT_NONCE_GENERATOR.generateKey();
+		String policyDirective = "script-src 'nonce-%s'".formatted(nonce);
+
+		ServerHttpResponse response = exchange.getResponse();
+		response.setStatusCode(HttpStatus.OK);
+		response.getHeaders().setContentType(MediaType.TEXT_HTML);
+		response.getHeaders().add(CONTENT_SECURITY_POLICY_HEADER, policyDirective);
+		return response.writeWith(createBuffer(exchange, location, nonce));
+	}
+
+	private Mono<DataBuffer> createBuffer(ServerWebExchange exchange, URI location, String nonce) {
+		byte[] bytes = createPage(location, nonce);
+		DataBufferFactory bufferFactory = exchange.getResponse().bufferFactory();
+		return Mono.just(bufferFactory.wrap(bytes));
+	}
+
+	private byte[] createPage(URI location, String nonce) {
+		UriComponentsBuilder uriComponentsBuilder = UriComponentsBuilder.fromUri(location);
+
+		StringBuilder hiddenInputsHtmlBuilder = new StringBuilder();
+		for (final Map.Entry<String, List<String>> entry : uriComponentsBuilder.build().getQueryParams().entrySet()) {
+			final String name = entry.getKey();
+			for (final String value : entry.getValue()) {
+				// @formatter:off
+				final String hiddenInput = HIDDEN_INPUT_TEMPLATE
+						.replace("{{name}}", HtmlUtils.htmlEscape(name))
+						.replace("{{value}}", HtmlUtils.htmlEscape(value));
+				// @formatter:on
+				hiddenInputsHtmlBuilder.append(hiddenInput.trim());
+			}
+		}
+		// @formatter:off
+		return REDIRECT_PAGE_TEMPLATE
+				.replace("{{action}}", HtmlUtils.htmlEscape(uriComponentsBuilder.query(null).build().toUriString()))
+				.replace("{{params}}", hiddenInputsHtmlBuilder.toString())
+				.replace("{{nonce}}", HtmlUtils.htmlEscape(nonce))
+				.getBytes(Charset.defaultCharset());
```

Comment:
I believe we will want to use `StandardCharsets.UTF_8` here instead since the html page declares that `charset` in a `<meta>` tag.